systemverilog-plugin/UhdmAst.cc

```diff
@@ -3430,8 +3430,37 @@ void UhdmAst::process_gen_scope()
 void UhdmAst::process_case()
 {
     current_node = make_ast_node(AST::AST_CASE);
+    auto cond_type = AST::AST_COND;
+    switch (vpi_get(vpiCaseType, obj_h)) {
+    case vpiCaseExact:
+        cond_type = AST::AST_COND;
+        break;
+    case vpiCaseX:
+        cond_type = AST::AST_CONDX;
+        break;
+    case vpiCaseZ:
+        cond_type = AST::AST_CONDZ;
+        break;
+    default: {
+        const uhdm_handle *const handle = (const uhdm_handle *)obj_h;
+        const UHDM::BaseClass *const object = (const UHDM::BaseClass *)handle->object;
+        report_error("%.*s:%d: Unknown case type", (int)object->VpiFile().length(), object->VpiFile().data(), object->VpiLineNo());
+    }
+    }
     visit_one_to_one({vpiCondition}, obj_h, [&](AST::AstNode *node) { current_node->children.push_back(node); });
-    visit_one_to_many({vpiCaseItem}, obj_h, [&](AST::AstNode *node) { current_node->children.push_back(node); });
+    visit_one_to_many({vpiCaseItem}, obj_h, [&](AST::AstNode *node) {
+        node->type = cond_type;
+        // TODO(krak): We are changing here Sx to Sz to match
+        // yosys read_verilog -sv output
+        if (cond_type == AST::AST_CONDX) {
+            for (unsigned i = 0; i < node->bits.size(); i++) {
+                if (node->bits[i] == RTLIL::State::Sx) {
+                    node->bits[i] = RTLIL::State::Sz;
+                }
+            }
+        }
```

Comment:
This doesn't seem to be true. I did check [your test][706], and a bit extended version (`case_xz.sv` below). `x` and `z` in `casex` and `casez` are replaced with `?` (`State::Sa`) by `read_verilog` before simplification. I believe this happens in [`const2ast`][const2ast]. Even though we use a copy of this file in the plugin, we're always passing `0` in `case_type` attribute, so we don't get the substitution.
`x`, and `z` are also converted into `?` [during simplification][cond_simplify], so after this step casex and casez constants from `read_verilog` and `read_systemverilog` are the same.

The code above should either convert `Sx` and `Sz` into `Sa` (for both `AST_CONDX` and `AST_CONDZ`), or not convert them at all.
I didn't check, but I assume calling `const2ast` with `case_type` set to either `x` or `z` just for those constants would require a bit more changes. If that's not true and can be done easily it would be a preferred solution.

[706]: https://github.com/chipsalliance/UHDM-integration-tests/pull/706
[const2ast]: https://github.com/YosysHQ/yosys/blob/417fadbefd0ed85846c2ba855457361999fcf18b/frontends/verilog/const2ast.cc#L109-L112
[cond_simplify]: https://github.com/YosysHQ/yosys/blob/417fadbefd0ed85846c2ba855457361999fcf18b/frontends/ast/simplify.cc#L1537-L1547

---

<details>
<summary><strong>The diff between AST (before simplification) from read_verilog and read_systemverilog used to parse <samp>case_xz.sv</samp>. Meaningless differences have been removed.</strong></summary>

```diff
--- read_verilog.out    2023-02-08 01:29:42.188144022 +0100
+++ read_systemverilog.out      2023-02-08 01:29:42.224143782 +0100
@@ -1,252 +1,252 @@
 Generating RTLIL representation for module `\top_za'.
 Dumping AST before simplification:
     AST_MODULE <case_xz.sv>  str='\top_za'
       AST_WIRE <case_xz.sv>  str='\s' input logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000010'(32) signed range=[31:0] int=2
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_WIRE <case_xz.sv>  str='\y' output logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) signed range=[31:0] int=1
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_ALWAYS <case_xz.sv>
         ATTR \always_comb:
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) range=[31:0] int=1
         AST_BLOCK <case_xz.sv>
           AST_CASE <case_xz.sv>
             AST_IDENTIFIER <case_xz.sv>  str='\s'
             AST_CONDZ <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='1??'(3) range=[2:0] int=4
+              AST_CONSTANT <case_xz.sv>  bits='1zz'(3) range=[2:0] int=4
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='11'(2) range=[1:0] int=3
             AST_CONDZ <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='01?'(3) range=[2:0] int=2
+              AST_CONSTANT <case_xz.sv>  bits='01z'(3) range=[2:0] int=2
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='10'(2) range=[1:0] int=2
             AST_CONDZ <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='001'(3) range=[2:0] int=1
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='01'(2) range=[1:0] int=1
             AST_CONDZ <case_xz.sv>
               AST_DEFAULT <case_xz.sv>
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='00'(2) range=[1:0]
 --- END OF AST DUMP ---
 Generating RTLIL representation for module `\top_zz'.
 Dumping AST before simplification:
     AST_MODULE <case_xz.sv>  str='\top_zz'
       AST_WIRE <case_xz.sv>  str='\s' input logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000010'(32) signed range=[31:0] int=2
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_WIRE <case_xz.sv>  str='\y' output logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) signed range=[31:0] int=1
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_ALWAYS <case_xz.sv>
         ATTR \always_comb:
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) range=[31:0] int=1
         AST_BLOCK <case_xz.sv>
           AST_CASE <case_xz.sv>
             AST_IDENTIFIER <case_xz.sv>  str='\s'
             AST_CONDZ <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='1??'(3) range=[2:0] int=4
+              AST_CONSTANT <case_xz.sv>  bits='1zz'(3) range=[2:0] int=4
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='11'(2) range=[1:0] int=3
             AST_CONDZ <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='01?'(3) range=[2:0] int=2
+              AST_CONSTANT <case_xz.sv>  bits='01z'(3) range=[2:0] int=2
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='10'(2) range=[1:0] int=2
             AST_CONDZ <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='001'(3) range=[2:0] int=1
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='01'(2) range=[1:0] int=1
             AST_CONDZ <case_xz.sv>
               AST_DEFAULT <case_xz.sv>
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='00'(2) range=[1:0]
 --- END OF AST DUMP ---
 Generating RTLIL representation for module `\top_zx'.
 Dumping AST before simplification:
     AST_MODULE <case_xz.sv>  str='\top_zx'
       AST_WIRE <case_xz.sv>  str='\s' input logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000010'(32) signed range=[31:0] int=2
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_WIRE <case_xz.sv>  str='\y' output logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) signed range=[31:0] int=1
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_ALWAYS <case_xz.sv>
         ATTR \always_comb:
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) range=[31:0] int=1
         AST_BLOCK <case_xz.sv>
           AST_CASE <case_xz.sv>
             AST_IDENTIFIER <case_xz.sv>  str='\s'
             AST_CONDZ <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='1xx'(3) range=[2:0] int=4
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='11'(2) range=[1:0] int=3
             AST_CONDZ <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='01x'(3) range=[2:0] int=2
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='10'(2) range=[1:0] int=2
             AST_CONDZ <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='001'(3) range=[2:0] int=1
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='01'(2) range=[1:0] int=1
             AST_CONDZ <case_xz.sv>
               AST_DEFAULT <case_xz.sv>
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='00'(2) range=[1:0]
 --- END OF AST DUMP ---
 Generating RTLIL representation for module `\top_xa'.
 Dumping AST before simplification:
     AST_MODULE <case_xz.sv>  str='\top_xa'
       AST_WIRE <case_xz.sv>  str='\s' input logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000010'(32) signed range=[31:0] int=2
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_WIRE <case_xz.sv>  str='\y' output logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) signed range=[31:0] int=1
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_ALWAYS <case_xz.sv>
         ATTR \always_comb:
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) range=[31:0] int=1
         AST_BLOCK <case_xz.sv>
           AST_CASE <case_xz.sv>
             AST_IDENTIFIER <case_xz.sv>  str='\s'
             AST_CONDX <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='1??'(3) range=[2:0] int=4
+              AST_CONSTANT <case_xz.sv>  bits='1zz'(3) range=[2:0] int=4
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='11'(2) range=[1:0] int=3
             AST_CONDX <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='01?'(3) range=[2:0] int=2
+              AST_CONSTANT <case_xz.sv>  bits='01z'(3) range=[2:0] int=2
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='10'(2) range=[1:0] int=2
             AST_CONDX <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='001'(3) range=[2:0] int=1
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='01'(2) range=[1:0] int=1
             AST_CONDX <case_xz.sv>
               AST_DEFAULT <case_xz.sv>
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='00'(2) range=[1:0]
 --- END OF AST DUMP ---
 Generating RTLIL representation for module `\top_xx'.
 Dumping AST before simplification:
     AST_MODULE <case_xz.sv>  str='\top_xx'
       AST_WIRE <case_xz.sv>  str='\s' input logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000010'(32) signed range=[31:0] int=2
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_WIRE <case_xz.sv>  str='\y' output logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) signed range=[31:0] int=1
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_ALWAYS <case_xz.sv>
         ATTR \always_comb:
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) range=[31:0] int=1
         AST_BLOCK <case_xz.sv>
           AST_CASE <case_xz.sv>
             AST_IDENTIFIER <case_xz.sv>  str='\s'
             AST_CONDX <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='1??'(3) range=[2:0] int=4
+              AST_CONSTANT <case_xz.sv>  bits='1xx'(3) range=[2:0] int=4
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='11'(2) range=[1:0] int=3
             AST_CONDX <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='01?'(3) range=[2:0] int=2
+              AST_CONSTANT <case_xz.sv>  bits='01x'(3) range=[2:0] int=2
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='10'(2) range=[1:0] int=2
             AST_CONDX <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='001'(3) range=[2:0] int=1
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='01'(2) range=[1:0] int=1
             AST_CONDX <case_xz.sv>
               AST_DEFAULT <case_xz.sv>
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='00'(2) range=[1:0]
 --- END OF AST DUMP ---
 Generating RTLIL representation for module `\top_xz'.
 Dumping AST before simplification:
     AST_MODULE <case_xz.sv>  str='\top_xz'
       AST_WIRE <case_xz.sv>  str='\s' input logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000010'(32) signed range=[31:0] int=2
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_WIRE <case_xz.sv>  str='\y' output logic port=_
         AST_RANGE <case_xz.sv>
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) signed range=[31:0] int=1
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000000'(32) signed range=[31:0]
       AST_ALWAYS <case_xz.sv>
         ATTR \always_comb:
           AST_CONSTANT <case_xz.sv>  bits='00000000000000000000000000000001'(32) range=[31:0] int=1
         AST_BLOCK <case_xz.sv>
           AST_CASE <case_xz.sv>
             AST_IDENTIFIER <case_xz.sv>  str='\s'
             AST_CONDX <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='1??'(3) range=[2:0] int=4
+              AST_CONSTANT <case_xz.sv>  bits='1zz'(3) range=[2:0] int=4
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='11'(2) range=[1:0] int=3
             AST_CONDX <case_xz.sv>
-              AST_CONSTANT <case_xz.sv>  bits='01?'(3) range=[2:0] int=2
+              AST_CONSTANT <case_xz.sv>  bits='01z'(3) range=[2:0] int=2
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='10'(2) range=[1:0] int=2
             AST_CONDX <case_xz.sv>
               AST_CONSTANT <case_xz.sv>  bits='001'(3) range=[2:0] int=1
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='01'(2) range=[1:0] int=1
             AST_CONDX <case_xz.sv>
               AST_DEFAULT <case_xz.sv>
               AST_BLOCK <case_xz.sv>
                 AST_ASSIGN_EQ <case_xz.sv>
                   AST_IDENTIFIER <case_xz.sv>  str='\y'
                   AST_CONSTANT <case_xz.sv>  bits='00'(2) range=[1:0]
 --- END OF AST DUMP ---
```

</details>

<details>
<summary><strong><samp>case_xz.sv</samp></strong></summary>

```systemverilog
module top_za(input logic [2:0] s, output logic [1:0] y);
always_comb
 casez (s)
  3'b1?? : y = 2'b11;
  3'b01? : y = 2'b10;
  3'b001 : y = 2'b01;
  default : y= 2'b00;
 endcase
endmodule

module top_zz(input logic [2:0] s, output logic [1:0] y);
always_comb
 casez (s)
  3'b1zz : y = 2'b11;
  3'b01z : y = 2'b10;
  3'b001 : y = 2'b01;
  default : y= 2'b00;
 endcase
endmodule

module top_zx(input logic [2:0] s, output logic [1:0] y);
always_comb
 casez (s)
  3'b1xx : y = 2'b11;
  3'b01x : y = 2'b10;
  3'b001 : y = 2'b01;
  default : y= 2'b00;
 endcase
endmodule

module top_xa(input logic [2:0] s, output logic [1:0] y);
always_comb
 casex (s)
  3'b1?? : y = 2'b11;
  3'b01? : y = 2'b10;
  3'b001 : y = 2'b01;
  default : y= 2'b00;
 endcase
endmodule

module top_xx(input logic [2:0] s, output logic [1:0] y);
always_comb
 casex (s)
  3'b1xx : y = 2'b11;
  3'b01x : y = 2'b10;
  3'b001 : y = 2'b01;
  default : y= 2'b00;
 endcase
endmodule

module top_xz(input logic [2:0] s, output logic [1:0] y);
always_comb
 casex (s)
  3'b1zz : y = 2'b11;
  3'b01z : y = 2'b10;
  3'b001 : y = 2'b01;
  default : y= 2'b00;
 endcase
endmodule
```

</details>

Reply:
You are right, we have case type in ``vpiParent`` of constant, so I've added correct ``case_type`` to ``const2ast``.
I've also updated UHDM-integration-tests with your extended ``case_xz.sv`` and now plugin output matches original frontend for all cases.